File: dungeons-1.21.11/src/main/java/fishmod/features/Reputation.java
package fishmod.features;

import fishmod.utils.HypixelApi;
import fishmod.utils.Misc;
import fishmod.utils.config.values.FishSettings;
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents;
import net.minecraft.client.MinecraftClient;
import net.minecraft.text.MutableText;
import net.minecraft.text.Text;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;
import java.util.function.Consumer;

/**
 * Crowd-sourced player reputation — a community "vouch / shitter list". Any player can be tagged by
 * UUID (they don't need the mod): {@code /vouch} for a good carry, {@code /shitter} to flag a
 * ditcher/scammer, {@code /unrep} to clear your tag. {@code /rep <player>} shows the aggregate, and
 * {@code /rep} with no name scans your current lobby and lists anyone who's been flagged.
 *
 * One vote per voter per target (re-voting overwrites), so counts can't be stuffed. Backed by the
 * worker /rep route (worker-reputation-snippet.js); commands report cleanly if it isn't deployed.
 */
public final class Reputation {

    private Reputation() {}

    // Lowercased IGNs of flagged (net-negative) players currently in the lobby, refreshed by a poll.
    private static final Set<String> flaggedIgns = ConcurrentHashMap.newKeySet();
    private static final int POLL_TICKS = 20 * 30; // ~30s between flag refreshes
    private static int pollTick = 0;

    /** Registers the background poll that keeps the in-lobby flagged set fresh (for the tab ✗ marker). */
    public static void init() {
        ClientTickEvents.END_CLIENT_TICK.register(mc -> {
            if (!FishSettings.repFlagsEnabled) { if (!flaggedIgns.isEmpty()) flaggedIgns.clear(); return; }
            if (++pollTick < POLL_TICKS) return;
            pollTick = 0;
            pollFlags(mc);
        });
    }

    private static void pollFlags(MinecraftClient mc) {
        if (mc.getNetworkHandler() == null) return;
        Map<String, String> uuidToName = new HashMap<>();
        for (var e : mc.getNetworkHandler().getPlayerList()) {
            var gp = e.getProfile();
            if (gp == null || gp.id() == null || gp.name() == null || gp.name().isBlank()) continue;
            uuidToName.put(gp.id().toString().replace("-", ""), gp.name());
        }
        if (uuidToName.isEmpty()) return;
        HypixelApi.fetchReps(uuidToName.keySet(), reps -> mc.execute(() -> {
            Set<String> next = new HashSet<>();
            for (var entry : reps.entrySet()) {
                HypixelApi.RepData rd = entry.getValue();
                if (rd.down() > rd.up()) {
                    String name = uuidToName.get(entry.getKey());
                    if (name != null) next.add(name.toLowerCase());
                }
            }
            flaggedIgns.clear();
            flaggedIgns.addAll(next);
        }));
    }

    /**
     * Appends a red ✘ to any flagged player's name found in an on-screen text (tab list / scoreboard).
     * Cheap: the flagged set is usually empty and only ever holds players in your current lobby.
     */
    public static Text decorateTab(Text text) {
        if (!FishSettings.repFlagsEnabled || text == null || flaggedIgns.isEmpty()) return text;
        String s = text.getString();
        if (s.isEmpty() || s.endsWith("✘")) return text;
        String lower = s.toLowerCase();
        for (String name : flaggedIgns) {
            if (lower.contains(name)) {
                MutableText out = text.copy();
                out.append(Text.literal(" §c✘"));
                return out;
            }
        }
        return text;
    }

    /** Resolve an IGN to a dashless UUID (cache first, then Mojang), then run {@code cb} (null on miss). */
    private static void withUuid(String name, Consumer<String> cb) {
        String cached = HypixelApi.getCachedUuid(name);
        if (cached != null) { cb.accept(cached.replace("-", "")); return; }
        HypixelApi.resolveUuidAsync(name, uuid -> cb.accept(uuid == null ? null : uuid.replace("-", "")));
    }

    /** Cast (or clear) the local player's vote on {@code name}. dir ∈ "up" | "down" | "none". */
    public static void vote(String name, String dir) {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.player == null) return;
        String voter = mc.player.getUuid().toString().replace("-", "");
        Misc.addChatMessage(Text.literal("§7[Rep] resolving §f" + name + "§7…"));
        withUuid(name, uuid -> {
            if (uuid == null) { mc.execute(() -> Misc.addChatMessage(Text.literal("§c[Rep] couldn't find player §f" + name))); return; }
            HypixelApi.voteRep(voter, uuid, name, dir, (up, down) -> mc.execute(() -> {
                if (up < 0) { Misc.addChatMessage(Text.literal("§c[Rep] vote failed — worker route not reachable.")); return; }
                String v = dir.equals("up") ? "§avouched" : dir.equals("down") ? "§cflagged" : "§7cleared vote on";
                Misc.addChatMessage(Text.literal("§7[Rep] " + v + " §f" + name + " §8— " + counts(up, down)));
            }));
        });
    }

    /** Print the aggregate reputation for a single player. */
    public static void lookup(String name) {
        MinecraftClient mc = MinecraftClient.getInstance();
        Misc.addChatMessage(Text.literal("§7[Rep] looking up §f" + name + "§7…"));
        withUuid(name, uuid -> {
            if (uuid == null) { mc.execute(() -> Misc.addChatMessage(Text.literal("§c[Rep] couldn't find player §f" + name))); return; }
            HypixelApi.fetchReps(Set.of(uuid), reps -> mc.execute(() -> {
                HypixelApi.RepData rd = reps.get(uuid);
                if (rd == null || (rd.up() == 0 && rd.down() == 0)) {
                    Misc.addChatMessage(Text.literal("§7[Rep] §f" + name + " §7has no reputation yet."));
                } else {
                    Misc.addChatMessage(Text.literal("§7[Rep] §f" + name + " §8— " + counts(rd.up(), rd.down()) + " " + verdict(rd)));
                }
            }));
        });
    }

    /** Scan the current lobby's tab list and list any flagged (net-negative) players. */
    public static void listNearby() {
        MinecraftClient mc = MinecraftClient.getInstance();
        if (mc.getNetworkHandler() == null) { Misc.addChatMessage(Text.literal("§c[Rep] Not in a world.")); return; }
        Map<String, String> uuidToName = new HashMap<>();
        for (var e : mc.getNetworkHandler().getPlayerList()) {
            var gp = e.getProfile();
            if (gp == null || gp.id() == null || gp.name() == null || gp.name().isBlank()) continue;
            uuidToName.put(gp.id().toString().replace("-", ""), gp.name());
        }
        if (uuidToName.isEmpty()) { Misc.addChatMessage(Text.literal("§7[Rep] No players found in tab.")); return; }
        Misc.addChatMessage(Text.literal("§7[Rep] scanning §f" + uuidToName.size() + " §7players…"));
        HypixelApi.fetchReps(uuidToName.keySet(), reps -> mc.execute(() -> {
            List<String> flagged = new ArrayList<>();
            for (var entry : reps.entrySet()) {
                HypixelApi.RepData rd = entry.getValue();
                if (rd.down() > rd.up()) {
                    String who = uuidToName.getOrDefault(entry.getKey(), rd.name());
                    flagged.add("§f" + who + " §8(" + counts(rd.up(), rd.down()) + "§8)" + (isShitter(rd) ? " §4§lSHITTER" : ""));
                }
            }
            if (flagged.isEmpty()) { Misc.addChatMessage(Text.literal("§a[Rep] No flagged players in this lobby.")); return; }
            Misc.addChatMessage(Text.literal("§c[Rep] §l" + flagged.size() + " flagged player(s) here:"));
            for (String line : flagged) Misc.addChatMessage(Text.literal("  " + line));
        }));
    }

    private static String counts(int up, int down) { return "§a+" + up + " §c-" + down; }

    private static boolean isShitter(HypixelApi.RepData rd) {
        return rd.down() >= 3 && rd.down() > rd.up() * 2;
    }

    private static String verdict(HypixelApi.RepData rd) {
        if (isShitter(rd)) return "§4§lSHITTER";
        if (rd.up() - rd.down() >= 3) return "§2§ltrusted";
        return "";
    }
}

File: dungeons-26.1.2/src/main/kotlin/fishmod/features/Reputation.kt
package fishmod.features

import fishmod.utils.HypixelApi
import fishmod.utils.Misc
import fishmod.utils.config.values.FishSettings
import net.fabricmc.fabric.api.client.event.lifecycle.v1.ClientTickEvents
import net.minecraft.client.Minecraft
import net.minecraft.network.chat.Component
import java.util.concurrent.ConcurrentHashMap

/**
 * Crowd-sourced player reputation — a community "vouch / shitter list". Any player can be tagged by
 * UUID (they don't need the mod): `/vouch` for a good carry, `/shitter` to flag a
 * ditcher/scammer, `/unrep` to clear your tag. `/rep <player>` shows the aggregate, and
 * `/rep` with no name scans your current lobby and lists anyone who's been flagged.
 *
 * One vote per voter per target (re-voting overwrites), so counts can't be stuffed. Backed by the
 * worker /rep route (worker-reputation-snippet.js); commands report cleanly if it isn't deployed.
 */
object Reputation {

    // Lowercased IGNs of flagged (net-negative) players currently in the lobby, refreshed by a poll.
    private val flaggedIgns: MutableSet<String> = ConcurrentHashMap.newKeySet()
    private const val POLL_TICKS = 20 * 30 // ~30s between flag refreshes
    private var pollTick = 0

    /** Registers the background poll that keeps the in-lobby flagged set fresh (for the tab ✗ marker). */
    @JvmStatic
    fun init() {
        ClientTickEvents.END_CLIENT_TICK.register { mc ->
            if (!FishSettings.repFlagsEnabled) {
                if (flaggedIgns.isNotEmpty()) flaggedIgns.clear()
                return@register
            }
            if (++pollTick < POLL_TICKS) return@register
            pollTick = 0
            pollFlags(mc)
        }
    }

    private fun pollFlags(mc: Minecraft) {
        val connection = mc.getConnection() ?: return
        val uuidToName = HashMap<String, String>()
        for (e in connection.getOnlinePlayers()) {
            val gp = e.getProfile()
            if (gp == null || gp.id() == null || gp.name() == null || gp.name().isBlank()) continue
            uuidToName[gp.id().toString().replace("-", "")] = gp.name()
        }
        if (uuidToName.isEmpty()) return
        HypixelApi.fetchReps(uuidToName.keys) { reps ->
            mc.execute {
                val next = HashSet<String>()
                for (entry in reps.entries) {
                    val rd: HypixelApi.RepData = entry.value
                    if (rd.down() > rd.up()) {
                        val name = uuidToName[entry.key]
                        if (name != null) next.add(name.lowercase())
                    }
                }
                flaggedIgns.clear()
                flaggedIgns.addAll(next)
            }
        }
    }

    /**
     * Appends a red ✘ to any flagged player's name found in an on-screen text (tab list / scoreboard).
     * Cheap: the flagged set is usually empty and only ever holds players in your current lobby.
     */
    @JvmStatic
    fun decorateTab(text: Component?): Component? {
        if (!FishSettings.repFlagsEnabled || text == null || flaggedIgns.isEmpty()) return text
        val s = text.getString()
        if (s.isEmpty() || s.endsWith("✘")) return text
        val lower = s.lowercase()
        for (name in flaggedIgns) {
            if (lower.contains(name)) {
                val out = text.copy()
                out.append(Component.literal(" §c✘"))
                return out
            }
        }
        return text
    }

    /** Resolve an IGN to a dashless UUID (cache first, then Mojang), then run `cb` (null on miss). */
    private fun withUuid(name: String, cb: (String?) -> Unit) {
        val cached = HypixelApi.getCachedUuid(name)
        if (cached != null) {
            cb(cached.replace("-", ""))
            return
        }
        HypixelApi.resolveUuidAsync(name) { uuid -> cb(uuid?.replace("-", "")) }
    }

    /** Cast (or clear) the local player's vote on `name`. dir ∈ "up" | "down" | "none". */
    @JvmStatic
    fun vote(name: String, dir: String) {
        val mc = Minecraft.getInstance()
        val player = mc.player ?: return
        val voter = player.getUUID().toString().replace("-", "")
        Misc.addChatMessage(Component.literal("§7[Rep] resolving §f$name§7…"))
        withUuid(name) { uuid ->
            if (uuid == null) {
                mc.execute { Misc.addChatMessage(Component.literal("§c[Rep] couldn't find player §f$name")) }
                return@withUuid
            }
            HypixelApi.voteRep(voter, uuid, name, dir) { up, down ->
                mc.execute {
                    if (up < 0) {
                        Misc.addChatMessage(Component.literal("§c[Rep] vote failed — worker route not reachable."))
                        return@execute
                    }
                    val v = if (dir == "up") "§avouched" else if (dir == "down") "§cflagged" else "§7cleared vote on"
                    Misc.addChatMessage(Component.literal("§7[Rep] $v §f$name §8— ${counts(up, down)}"))
                }
            }
        }
    }

    /** Print the aggregate reputation for a single player. */
    @JvmStatic
    fun lookup(name: String) {
        val mc = Minecraft.getInstance()
        Misc.addChatMessage(Component.literal("§7[Rep] looking up §f$name§7…"))
        withUuid(name) { uuid ->
            if (uuid == null) {
                mc.execute { Misc.addChatMessage(Component.literal("§c[Rep] couldn't find player §f$name")) }
                return@withUuid
            }
            HypixelApi.fetchReps(setOf(uuid)) { reps ->
                mc.execute {
                    val rd = reps[uuid]
                    if (rd == null || (rd.up() == 0 && rd.down() == 0)) {
                        Misc.addChatMessage(Component.literal("§7[Rep] §f$name §7has no reputation yet."))
                    } else {
                        Misc.addChatMessage(Component.literal("§7[Rep] §f$name §8— ${counts(rd.up(), rd.down())} ${verdict(rd)}"))
                    }
                }
            }
        }
    }

    /** Scan the current lobby's tab list and list any flagged (net-negative) players. */
    @JvmStatic
    fun listNearby() {
        val mc = Minecraft.getInstance()
        val connection = mc.getConnection()
        if (connection == null) {
            Misc.addChatMessage(Component.literal("§c[Rep] Not in a world."))
            return
        }
        val uuidToName = HashMap<String, String>()
        for (e in connection.getOnlinePlayers()) {
            val gp = e.getProfile()
            if (gp == null || gp.id() == null || gp.name() == null || gp.name().isBlank()) continue
            uuidToName[gp.id().toString().replace("-", "")] = gp.name()
        }
        if (uuidToName.isEmpty()) {
            Misc.addChatMessage(Component.literal("§7[Rep] No players found in tab."))
            return
        }
        Misc.addChatMessage(Component.literal("§7[Rep] scanning §f${uuidToName.size} §7players…"))
        HypixelApi.fetchReps(uuidToName.keys) { reps ->
            mc.execute {
                val flagged = ArrayList<String>()
                for (entry in reps.entries) {
                    val rd: HypixelApi.RepData = entry.value
                    if (rd.down() > rd.up()) {
                        val who = uuidToName.getOrDefault(entry.key, rd.name())
                        flagged.add("§f$who §8(${counts(rd.up(), rd.down())}§8)" + (if (isShitter(rd)) " §4§lSHITTER" else ""))
                    }
                }
                if (flagged.isEmpty()) {
                    Misc.addChatMessage(Component.literal("§a[Rep] No flagged players in this lobby."))
                    return@execute
                }
                Misc.addChatMessage(Component.literal("§c[Rep] §l${flagged.size} flagged player(s) here:"))
                for (line in flagged) Misc.addChatMessage(Component.literal("  $line"))
            }
        }
    }

    private fun counts(up: Int, down: Int): String = "§a+$up §c-$down"

    private fun isShitter(rd: HypixelApi.RepData): Boolean = rd.down() >= 3 && rd.down() > rd.up() * 2

    private fun verdict(rd: HypixelApi.RepData): String {
        if (isShitter(rd)) return "§4§lSHITTER"
        if (rd.up() - rd.down() >= 3) return "§2§ltrusted"
        return ""
    }
}
